File: Recursion/PalindromicDecomposition.java
public class PalindromicDecomposition {
    boolean isPalindrome(char[] s, int begin, int end) {
        return begin >= end || s[begin] == s[end] && isPalindrome(s, begin + 1, end - 1);
    }

    void palindromicDecompositionRecursion(char[] s, int begin, char[] decomposition, int write) {
        if (begin == s.length) {
            for (int i = 0; i < write; ++i)
                System.out.print(decomposition[i]);
            System.out.println();
            return;
        }

        for(int end = begin; end < s.length; ++end) {
            if (isPalindrome(s, begin, end)) {
                int w = write;
                for(int j = begin; j <= end ; ++j)
                    decomposition[w++] = s[j];
                decomposition[w++] = '|';
                palindromicDecompositionRecursion(s, end + 1, decomposition, w);
            }
        }
    }

    void decompose(String s) {
        char[] decomposition = new char[2 * s.length()];
        palindromicDecompositionRecursion(s.toCharArray(), 0, decomposition, 0);
    }

    public static void main(String[] args) {
        PalindromicDecomposition pD = new PalindromicDecomposition();
        String s = args.length > 0 ? args[0] : "abacddcaba";
        pD.decompose(s);
    }
}

File: Recursion/palindromic decomposition/palindromic_decomposition.c
/******************************************************************************
 *
 *  Palindromic Decomposition
 *
 *****************************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int isPalindrome(const char *begin, const char *end) {
    if (begin >= end)
        return 1;
    return *begin == *end && isPalindrome(begin + 1, end - 1); 
}

void palindromicDecompositionRecursion(const char *string, char *decomposition, char *write) {
    if (*string == '\0') {
        *write = '\0';
        printf("%s\n", decomposition);
        return;
    }
    
    for(const char *end = string; *end != '\0'; end++) {
        if (isPalindrome(string, end)) {
            char *w = write;
            for(const char *j = string; j <= end ; ++j)
                *w++ = *j;
            *w++ = '|';
            palindromicDecompositionRecursion(end + 1, decomposition, w);
        }    
    }
}

void palindromicDecomposition(const char *string) {
    char *decomposition = malloc(2 * strlen(string) + 1);
    palindromicDecompositionRecursion(string, decomposition, decomposition);
    free(decomposition);
}

int main(int argc, char **argv) {
    palindromicDecomposition(argv[1]);
    return 0;
} 
